query/stdlib/influxdata/influxdb/from.go

```diff
@@ -3,23 +3,576 @@ package influxdb
+	fromSpec := fromNode.ProcedureSpec().(*FromProcedureSpec)
+
+	if fromSpec.LimitSet && fromSpec.PointsLimit == -1 {
+		return distinctNode, false, nil
+	}
+
+	groupStar := !fromSpec.GroupingSet && distinctSpec.Column != execute.DefaultValueColLabel && distinctSpec.Column != execute.DefaultTimeColLabel
+	groupByColumn := fromSpec.GroupingSet && len(fromSpec.GroupKeys) > 0 &&
+		((fromSpec.GroupMode == flux.GroupModeBy && execute.ContainsStr(fromSpec.GroupKeys, distinctSpec.Column)) ||
+			(fromSpec.GroupMode == flux.GroupModeExcept && !execute.ContainsStr(fromSpec.GroupKeys, distinctSpec.Column)))
+	if groupStar || groupByColumn {
+		newFromSpec := fromSpec.Copy().(*FromProcedureSpec)
+		newFromSpec.LimitSet = true
+		newFromSpec.PointsLimit = -1
+		if err := fromNode.ReplaceSpec(newFromSpec); err != nil {
+			return nil, false, err
+		}
+		return distinctNode, true, nil
+	}
+
+	return distinctNode, false, nil
+}
+
+type MergeFromGroupRule struct {
+}
+
+func (MergeFromGroupRule) Name() string {
+	return "MergeFromGroupRule"
+}
+
+func (MergeFromGroupRule) Pattern() plan.Pattern {
+	return plan.Pat(universe.GroupKind, plan.Pat(FromKind))
+}
+
+func (MergeFromGroupRule) Rewrite(groupNode plan.PlanNode) (plan.PlanNode, bool, error) {
+	fromNode := groupNode.Predecessors()[0]
+	groupSpec := groupNode.ProcedureSpec().(*universe.GroupProcedureSpec)
+	fromSpec := fromNode.ProcedureSpec().(*FromProcedureSpec)
+
+	if fromSpec.GroupingSet ||
+		fromSpec.LimitSet ||
+		groupSpec.GroupMode != flux.GroupModeBy {
+		return groupNode, false, nil
+	}
+
+	for _, c := range groupSpec.GroupKeys {
+		// Storage can only do grouping over tag keys.
+		// Note: _start and _stop are okay, since storage is always implicitly grouping by them anyway.
+		if c == execute.DefaultTimeColLabel || c == execute.DefaultValueColLabel {
+			return groupNode, false, nil
+		}
+	}
+
+	newFromSpec := fromSpec.Copy().(*FromProcedureSpec)
+	newFromSpec.GroupingSet = true
+	newFromSpec.GroupMode = groupSpec.GroupMode
+	newFromSpec.GroupKeys = groupSpec.GroupKeys
+	merged, err := plan.MergePhysicalPlanNodes(groupNode, fromNode, newFromSpec)
+	if err != nil {
+		return nil, false, err
+	}
+	return merged, true, nil
+}
+
+type FromKeysRule struct {
+}
+
+func (FromKeysRule) Name() string {
+	return "FromKeysRule"
+}
+
+func (FromKeysRule) Pattern() plan.Pattern {
+	return plan.Pat(universe.KeysKind, plan.Pat(FromKind))
+}
+
+func (FromKeysRule) Rewrite(keysNode plan.PlanNode) (plan.PlanNode, bool, error) {
+	fromNode := keysNode.Predecessors()[0]
+	fromSpec := fromNode.ProcedureSpec().(*FromProcedureSpec)
+
+	if fromSpec.LimitSet && fromSpec.PointsLimit == -1 {
+		return keysNode, false, nil
+	}
+
+	newFromSpec := fromSpec.Copy().(*FromProcedureSpec)
+	newFromSpec.LimitSet = true
+	newFromSpec.PointsLimit = -1
+
+	if err := fromNode.ReplaceSpec(newFromSpec); err != nil {
+		return nil, false, err
+	}
+
+	return keysNode, true, nil
```

Comment:
I'm assuming that you only had to make minimal changes to move this code here so I'm not looking at it thoroughly... let me know if I should look at it closer...

Reply:
Exactly, I only changed imports.